### PR TITLE
find macro-feature occurrences within macro args

### LINKED ISSUE #22234: `feature_gate::MacroVisitor` misses macro invocations nested in macro invocations
Today, this program is correctly rejected:
```rust
fn main() {
    unsafe {
        let x = asm!("");
        println!("hi {:?}", x);
    }
}
```
due to a missing `#![feature(asm)]`

Unfortunately, this variant program is accepted by the compiler:
```rust
fn main() {
    unsafe {
        println!("hi {:?}", asm!(""));
    }
}
```

The reason (I think) is that the `fn visit_mac` in `feature_gate::MacroVisitor` needs to recursively invoke the visitor on the expression inputs to the macro.

### PR DESCRIPTION
**Pull request by pnkfelix**
In feature_gate::MacroVisitor, find feature occurrences within macro arguments.

Fix #22234

### CODE REVIEW

**Comment by huonw** [comment]
This doesn't quite catch *everything*, e.g.

```rust
macro_rules! foo {
     ($x: ident) => {
         $x!("")
     }
}

foo!(asm)
```

is valid, AFAIK.

**Comment by pnkfelix** [comment]
Note: I am not actually thrilled with the approach I used here; it "works", and we may want to include it nonetheless.

But I probably need (for other checks, namely the check for placement-`in` and overloaded-`box`) to take a different approach.  E.g. inject a marker (saying that the feature had been used) into the generated AST, as outlined in my comment here: https://github.com/rust-lang/rust/issues/22234#issuecomment-74136635

(And if I do the latter, then the fix given here has a lower priority.)

**Comment by pnkfelix** [comment]
@huonw yes, it doesn't catch that.  The main thing it catches that my "inject-a-marker" approach would not catch is a scenario where the user has a macro that discards its input.  (And that probably is not so important to catch.)

The other thing I have been wondering is if there are legitimate macros that this check would erroneously reject.

**Comment by huonw** [comment]
Oh, I just thought a much better way to do this sort of feature gating: just register the existence of the various macro-related `feature(...)`s in a struct like the `Features` one that already exists, and have each macro do the feature gating check internally. That is, change e.g. `asm`s definition to start like:

```rust
pub fn expand_quote_tokens<'cx>(cx: &'cx mut ExtCtxt,
                                sp: Span,
                                tts: &[ast::TokenTree])
                                -> Box<base::MacResult+'cx> {

    cx.sess.macro_features.gate(/* name of the feature */ "asm", /* name of the macro */ "asm");
```

where the `macro_features`s field is storing what I mentioned above.

This is then "perfect": any expansion of such a macro will be caught.

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -225,29 +226,44 @@ struct MacroVisitor<'a> {
     context: &'a Context<'a>
 }
 
+impl<'a> MacroVisitor<'a> {
+    fn check_ident(&self, id: ast::Ident, span: Span) {
+        for &(s, readable) in &[("asm", "inline assembly"),
+                               ("log_syntax", "`log_syntax!`"),
```
Comment (nikomatsakis) [inline src/libsyntax/feature_gate.rs]: nit: indentation

**Comment by nikomatsakis** [comment]
@huonw huh. I always thought of `foo!` as a single token. I'm...not thrilled that this is not the case. (In particular, it seems to make possible future keyword and macro overlap mildly more troublesome.)

**Comment by nikomatsakis** [comment]
So, this code seems fine, though the technique @huonw suggests seems more robust.

**Comment by pnkfelix** [comment]
@nikomatsakis Indeed, I am in the process of implementing a feature much like the one @huonw suggested.

(We may still want to do something like this PR, because occurrences of the macro that are hidden under `#[cfg(..)]` and also under a macro argument will not be caught.  But then again, that seems like a sufficiently remote corner case that we can live with it.)

**Comment by pnkfelix** [comment]
(hein, I'll just close this PR; I think I have something close to working for the other approach, and I am not really convinced by the corner case I outlined above.)

**Pull request closed without merging**